### ISSUE TITLE
hal_ee24xx: potential infinite loop caused by __fitted_write when eeprom_write returns anything other than MSG_OK

### ISSUE THREAD

**Issue by apmorton** [ChibiOS/ChibiOS-Contrib#138]
in the case where `eeprom_write` fails repeatedly `__fitted_write` will never increment the value given to its `written` parameter.
https://github.com/ChibiOS/ChibiOS-Contrib/blob/9d74dd2661b80a5ae8598591f0251b197cc51756/os/hal/src/hal_ee24xx.c#L207-L212

which in turn causes an infinite loop in `write` as shown below
https://github.com/ChibiOS/ChibiOS-Contrib/blob/9d74dd2661b80a5ae8598591f0251b197cc51756/os/hal/src/hal_ee24xx.c#L262-L266

`eeprom_write` simply passes on the return value from `i2cMasterTransmitTimeout`, which can return `MSG_TIMEOUT` if there is a timeout or `MSG_RESET` if an I2C bus error occurs.

The result is that if you have a mis-behaving eeprom, the software will deadlock.

In addition, if some transient error condition occurs, it would go unnoticed and result in very different data being written than expected - as well as read past end of buffer errors on `bp` since `bp` is incremented by `len` regardless of whether `written` gets incremented.

In short, the `write` method of the ee24xx driver is incredibly unreliable and has many bugs.
Quickly looking at the ee25xx driver (although I have not used it), the infinite loop issue was fixed, but the transient error issue still exists since the return value of `__fitted_write` is not checked every time it is called.

**Comment by fpoussin** [comment]
Thanks for reporting this issue.
Since you found out about it, do you have any fixes to merge?
